File: 1250-longest-common-subsequence/longest-common-subsequence.c
//Since the forward loop is more efficient, the recurrence used
// here is different from the recursive version
#pragma GCC optimize("O3", "unroll-loops")
int longestCommonSubsequence(char* text1, char* text2) {
    int n=strlen(text1), m=strlen(text2);
    unsigned* dp[2];
    
    for (register int i = 0; i <= 1; i++) {
        dp[i] = (unsigned*)malloc((m+1)*sizeof(unsigned));
        memset(dp[i], 0, (m+1)*sizeof(unsigned));
    }

    for (register int i=1; i <= n; i++) {
        for (register int j = 1; j <= m; j++) {
            if (text1[i-1]==text2[j-1]) 
                dp[i&1][j]=1+dp[(i-1)&1][j-1];
            else 
                dp[i&1][j]=(dp[(i-1)&1][j]>dp[i&1][j-1])?dp[(i-1)&1][j]:dp[i&1][j-1];
        }
    }

    int ans=dp[n&1][m];

    // Free allocated memory
    for (register int i = 0; i <= 1; i++) 
        free(dp[i]);

    return ans;
}

File: 1250-longest-common-subsequence/longest-common-subsequence.kt
class Solution {
    fun longestCommonSubsequence(text1: String, text2: String): Int =
        (if (text1.length < text2.length) Pair(text2, text1) else Pair(text1, text2)).let { (t1, t2) ->
            (1..t1.length).fold(Array(2) { IntArray(t2.length + 1) }) { m, i ->
                (1..t2.length).forEach { j -> 
                    m[i % 2][j] = if (t1[i - 1] == t2[j - 1]) m[(i - 1) % 2][j - 1] + 1 else maxOf(m[(i - 1) % 2][j], m[i % 2][j - 1])
                }
            m }[t1.length % 2].last() }
}